File: hyper-frames/tools/studio/discovery/src/studio_discovery/ids.py
"""Stable raw_signal ids — MUST match server/lib/ids.ts byte-for-byte."""

from __future__ import annotations

import hashlib

# unit separator (0x1f) — keeps source/external_id boundaries unambiguous
_US = "\x1f"


def raw_signal_id(source: str, external_id: str) -> str:
    """sha1(source + 0x1f + external_id), first 16 hex chars."""
    return hashlib.sha1(f"{source}{_US}{external_id}".encode()).hexdigest()[:16]

File: hyper-frames/tools/studio/server/lib/ids.ts
import { createHash } from "node:crypto";

/** Unit separator (0x1f) — keeps source/externalId boundaries unambiguous. */
const US = String.fromCharCode(0x1f);

/**
 * Stable id for a raw_signal row. MUST stay byte-for-byte identical to the
 * Python `studio_discovery.ids.raw_signal_id` so the two writers dedup against
 * the same keyspace: sha1(source + 0x1f + externalId), first 16 hex chars.
 */
export function rawSignalId(source: string, externalId: string): string {
  return createHash("sha1")
    .update(source + US + externalId)
    .digest("hex")
    .slice(0, 16);
}
